Rewrite logp graph before taking the gradient

dehorsley opened the pull request:
With pymc-devs/pytensor#303 merged, `cholesky(L.dot(L.T))` will be rewritten to `L` if `L.tag.lower_triangular` is set. This change adds these where appropriate. This is important for #6717, however more work is likely required to improve the gradient in such cases.

Some rough benchmarks of computing logp and its grad on the initial point of following model.

<details>

```
n=1000
with pm.Model() as m:
    chol, corr, sigmas = pm.LKJCholeskyCov('cov', n=n, eta=1, sd_dist=pm.HalfNormal.dist())
    pm.MvNormal('y', mu=np.zeros(n), chol=chol, observed=np.ones((1000, n)))

```

</details>


|  | C backend |  JAX backend |
|-|---|-----|
|before | 292 ms ± 28.6 ms | 285 ms ± 28.9 ms  | 
|after | 260 ms ± 19.9 ms | 107 ms ± 969 µs |

The major difference in the "after" between C and JAX backends is that JAX is computing the grad after the rewrite is applied. As mentioned in #6717, this is probably a good motivator for performing some kind of rewrite before computing the gradient.

## Maintenance
- improve performance of MvNormal logp with Cholesky factor.



<!-- readthedocs-preview pymc start -->
----
:books: Documentation preview :books:: https://pymc--6736.org.readthedocs.build/en/6736/

<!-- readthedocs-preview pymc end -->

ricardoV94 commented on pymc/distributions/multivariate.py:
```diff
@@ -1456,6 +1459,8 @@ def dist(cls, eta, n, sd_dist, *, compute_corr=True, **kwargs):
     def helper_deterministics(cls, n, packed_chol):
         chol = pm.expand_packed_triangular(n, packed_chol, lower=True)
+        # tag as lower triangular to enable pytensor rewries
+        chol.tag.lower_triangular = True
```
Can this be done in the `expand_packed_triangular`?

dehorsley replied:
good idea!

ricardoV94 replied:
We can remove it from here right?

ricardoV94 commented on pymc/logprob/rewriting.py:
```diff
@@ -351,3 +351,18 @@ def construct_ir_fgraph(
+pregrad_logprob_rewrites_db.register("stabilize", optdb.query("+stabilize"), "basic")
+
+
+def rewrite_pregrad(graph, rewriter=None):
```
I would maybe put this in `pytensorf`, we already have some custom rewrite logic going on there

ricardoV94 commented on pymc/model.py:
```diff
@@ -308,6 +309,8 @@ class ValueGradFunction:
         back from the array dtype to the variable dtype.
     compute_grads: bool, default=True
         If False, return only the logp, not the gradient.
+    pregrad_rewrite: bool, default=True
```
I would be opinionated and always run it (and not give the flag to users)?

ricardoV94 commented on pymc/logprob/rewriting.py:
```diff
@@ -351,3 +351,18 @@ def construct_ir_fgraph(
+pregrad_logprob_rewrites_db = SequenceDB()
+pregrad_logprob_rewrites_db.name = "pregrad_logprob_rewrites_db"
+pregrad_logprob_rewrites_db.register("pre-canonicalize", optdb.query("+canonicalize"), "basic")
+pregrad_logprob_rewrites_db.register("stabilize", optdb.query("+stabilize"), "basic")
```
I wonder if this might be an overkill, should we start with only cherry picked rewrites?

dehorsley replied:
@ricardoV94 I'm not sure what we should do here. What kind of inclusion criteria are you thinking? I imagine anything that simplifies or improves the stability of the logp graph would be nice?

ricardoV94 replied:
I was thinking of starting with just the things we know about. The reason for this is that we might want to use this as the "base" for pre-grad rewrites in PyTensor if/when we decide to make them lazy. This would be a good way to start learning what rewrites seem useful for gradients...  Probably everything in `canonicalize` makes sense as that makes the graphs "homogenous", but everything in `stabilize` might be an overkill or even counterproductive. `stabilizing` the base graph may make the gradient one more complex.

On another angle, I worry that the extra rewrites could make compilation much slower.

Could you perhaps benchmark how long does it take to obtain `dlogp` and, separately, to compile it for 2-3 different models?

The compilation should be checked with and without having already cached results. You can erase the pytensor cache with `pytensor-cache purge` or something similar.

ricardoV94 replied:
Yeah, I think this would be the easiest. If the rewrite time is not too affected by including all of `stabilize` vs not (just to test more easily), then I am okay with the current approach.

ricardoV94 reviewed (commented):
Looks pretty good, just wondering whether we should be more targeted in the rewrites we include

dehorsley committed:
Apply rewrites to logp before apply grad

This adds safe rewrites to logp before the grad operator is applied.
This is motivated by #6717, where expensive `cholesky(L.dot(L.T))`
operations are removed. If these remain in the logp graph when the grad
is taken, the resulting dlogp graph contains unnecessary operations.
However this may improve the stability and performance of grad logp
in other situation.

dehorsley committed:
Add lower triangular tags to allow chol rewrites

Since pymc-devs/pytensor#303, `cholesky(L.dot(L.T))` will rewritten to L
if `L.tag.lower_triangular=True`. This change adds these where
appropriate.

Fixes #6717.

ricardoV94 commented on pymc/distributions/multivariate.py:
```diff
@@ -132,6 +132,9 @@ def quaddist_matrix(cov=None, chol=None, tau=None, lower=True, *args, **kwargs):
         if chol.ndim != 2:
             raise ValueError("chol must be two dimensional.")
+
+        # tag as lower triangular to enable pytensor rewriets of chol(l.l') -> l
```
```suggestion
        # tag as lower triangular to enable pytensor rewrites of chol(l.l') -> l
```

ricardoV94 commented on pymc/math.py:
```diff
@@ -443,11 +443,17 @@ def expand_packed_triangular(n, packed, lower=True, diagonal_only=False):
         out = pt.zeros((n, n), dtype=pytensor.config.floatX)
         idxs = np.tril_indices(n)
-        return pt.set_subtensor(out[idxs], packed)
+        # tag as lower triangular to enable pytensor rewries
```
```suggestion
        # tag as lower triangular to enable pytensor rewrites
```

ricardoV94 commented on pymc/math.py:
```diff
@@ -443,11 +443,17 @@ def expand_packed_triangular(n, packed, lower=True, diagonal_only=False):
         out = pt.zeros((n, n), dtype=pytensor.config.floatX)
         idxs = np.triu_indices(n)
-        return pt.set_subtensor(out[idxs], packed)
+        # tag as upper triangular to enable pytensor rewries
```
```suggestion
        # tag as upper triangular to enable pytensor rewrites
```

ricardoV94 commented on tests/test_pytensorf.py:
```diff
@@ -878,3 +879,31 @@ def replacement_fn(var, replacements):
+
+def test_mvnormal_no_cholesky_op():
+    """
+    Test MvNormal likelyhood when using Cholesky factor parameterization does not unnecessarily
```
```suggestion
    Test MvNormal likelihood when using Cholesky factor parameterization does not unnecessarily
```

ricardoV94 reviewed (commented):
Some small typos.

Does the last PyTensor already include the rewrites?

ricardoV94 commented on pymc/pytensorf.py:
```diff
@@ -1228,3 +1230,18 @@ def constant_fold(
+pregrad_logprob_rewrites_db.register("stabilize", optdb.query("+stabilize"), "basic")
+
+
+def rewrite_pregrad(graph, rewriter=None):
```
Can we use this helper? 

https://github.com/pymc-devs/pytensor/blob/60bc3688f67b2bcb0fa4f833d2d7e3f4b2671bb7/pytensor/graph/rewriting/utils.py#L20

Also, any reason why you created a custom rewrite db, instead of just doing a query on the global `optdb`? I think the second should be fine? That's what using the helper would do, by just specifying `include`

dehorsley replied:
I created a custom rewrite db, as I was expecting to be more precise about what we wanted and followed the logprob module. Going with the helper looks better to me! I fix this up in the next day or so.

ricardoV94 replied:
I see, the logprob submodule is special in that those rewrites are not general, but exclusive to the rv-logp conversion. Here we're thinking about general rewrites so the `optdb` database makes sense (at least initially).

ricardoV94 reviewed (changes requested):
Just a last-minute doubt about the helper. WDYT?
